lib/sudo-system/src/time.rs

```diff
@@ -0,0 +1,192 @@
+use std::{
+    io::{Read, Write},
+    mem::MaybeUninit,
+    ops::{Add, Sub},
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SystemTime {
+    secs: i64,
+    nsecs: i64,
+}
+
+impl SystemTime {
+    pub(crate) fn new(secs: i64, nsecs: i64) -> SystemTime {
+        let mut secs = secs + nsecs / 1_000_000_000;
+        let mut nsecs = nsecs % 1_000_000_000;
+
+        if secs > 0 && nsecs < 0 {
+            secs -= 1;
+            nsecs += 1_000_000_000;
+        } else if secs < 0 && nsecs > 0 {
+            secs += 1;
+            nsecs -= 1_000_000_000;
+        }
+
+        SystemTime { secs, nsecs }
+    }
+
+    pub fn now() -> std::io::Result<SystemTime> {
+        let mut spec = MaybeUninit::<libc::timespec>::uninit();
+        sudo_cutils::cerr(unsafe { libc::clock_gettime(libc::CLOCK_BOOTTIME, spec.as_mut_ptr()) })?;
+        let spec = unsafe { spec.assume_init() };
+        Ok(spec.into())
+    }
+
+    pub(crate) fn encode(&self, target: &mut impl Write) -> std::io::Result<()> {
+        let secs = self.secs.to_ne_bytes();
+        let nsecs = self.nsecs.to_ne_bytes();
+        target.write_all(&secs)?;
+        target.write_all(&nsecs)?;
+        Ok(())
+    }
+
+    pub(crate) fn decode(from: &mut impl Read) -> std::io::Result<SystemTime> {
+        let mut sec_bytes = [0; 8];
+        let mut nsec_bytes = [0; 8];
+
+        from.read_exact(&mut sec_bytes)?;
+        from.read_exact(&mut nsec_bytes)?;
+
+        Ok(SystemTime::new(
+            i64::from_ne_bytes(sec_bytes),
+            i64::from_ne_bytes(nsec_bytes),
+        ))
+    }
+}
+
+impl Sub<SystemTime> for SystemTime {
+    type Output = Duration;
+
+    fn sub(self, rhs: SystemTime) -> Self::Output {
+        Duration::new(self.secs - rhs.secs, self.nsecs - rhs.nsecs)
+    }
+}
+
+impl Add<Duration> for SystemTime {
+    type Output = SystemTime;
+
+    fn add(self, rhs: Duration) -> Self::Output {
+        SystemTime::new(self.secs + rhs.secs, self.nsecs + rhs.nsecs)
+    }
+}
+
+impl Sub<Duration> for SystemTime {
+    type Output = SystemTime;
+
+    fn sub(self, rhs: Duration) -> Self::Output {
+        SystemTime::new(self.secs - rhs.secs, self.nsecs - rhs.nsecs)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub struct Duration {
+    secs: i64,
+    nsecs: i64,
+}
+
+impl Duration {
+    pub fn new(secs: i64, nsecs: i64) -> Duration {
+        let mut secs = secs + nsecs / 1_000_000_000;
+        let mut nsecs = nsecs % 1_000_000_000;
+
+        if secs > 0 && nsecs < 0 {
+            secs -= 1;
+            nsecs += 1_000_000_000;
+        } else if secs < 0 && nsecs > 0 {
+            secs += 1;
+            nsecs -= 1_000_000_000;
+        }
```

Comment:
(Same comment as for `SystemTime::new`, except here they are maybe not dead code.)
```suggestion
        let secs = secs + nsecs.div_euclid(1_000_000_000);
        let nsecs = nsecs.rem_euclid(1_000_000_000);
```